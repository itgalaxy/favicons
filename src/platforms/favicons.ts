import { FaviconHtmlElement } from "../index";
import { FaviconOptions, NamedIconOptions } from "../config/defaults";
import { transparentIcon, transparentIcons } from "../config/icons";
import { OptionalMixin, Platform, uniformIconOptions } from "./base";

const ICONS_OPTIONS: (NamedIconOptions & OptionalMixin)[] = [
  { name: "favicon.ico", ...transparentIcons(16, 24, 32, 48, 64) },
  { name: "favicon-16x16.png", ...transparentIcon(16) },
  { name: "favicon-32x32.png", ...transparentIcon(32) },
  { name: "favicon-48x48.png", ...transparentIcon(48) },
  { name: "favicon.svg", ...transparentIcon(1024), optional: true }, // arbitrary size. if more than one svg source is given, the closest to this size will be picked.
];

export class FaviconsPlatform extends Platform {
  constructor(options: FaviconOptions) {
    super(
      options,
      uniformIconOptions(options, options.icons.favicons, ICONS_OPTIONS),
    );
  }

  override async createHtml(): Promise<FaviconHtmlElement[]> {
    return this.iconOptions.map(({ name, ...options }) => {
      if (name.endsWith(".ico")) {
        // prettier-ignore
        return `<link rel="icon" type="image/x-icon" href="${this.cacheBusting(this.relative(name))}">`;
      } else if (name.endsWith(".svg")) {
        // prettier-ignore
        return `<link rel="icon" type="image/svg+xml" href="${this.cacheBusting(this.relative(name))}">`;
      }

      const { width, height } = options.sizes[0];

      // prettier-ignore
      return `<link rel="icon" type="image/png" sizes="${width}x${height}" href="${this.cacheBusting(this.relative(name))}">`;
    });
  }
}
