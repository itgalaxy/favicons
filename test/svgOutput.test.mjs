import favicons from "../src";
import { logo_png, logo_svg } from "./util";

test("should reuse svg as a favicon", async () => {
  expect.assertions(1);

  const result = await favicons(logo_svg, {
    icons: {
      favicons: ["favicon.svg"],
      android: false,
      appleIcon: false,
      appleStartup: false,
      windows: false,
      yandex: false,
    },
  });

  await expect(result).toMatchFaviconsSnapshot();
});

test("should generate svg favicon", async () => {
  expect.assertions(2);

  const result = await favicons(logo_png, {
    icons: {
      favicons: ["favicon.svg"],
      android: false,
      appleIcon: false,
      appleStartup: false,
      windows: false,
      yandex: false,
    },
  });

  const svg = result.images[0].contents;
  const deterministicSvg = Buffer.from(
    svg.toString().replace(/png;base64,[^"]+/, ""),
  );
  await expect({
    ...result,
    images: [
      {
        ...result.images[0],
        contents: deterministicSvg,
      },
    ],
  }).toMatchFaviconsSnapshot();

  const roundtripResult = await favicons(svg, {
    icons: {
      android: false,
      appleIcon: false,
      appleStartup: false,
      windows: false,
      yandex: false,
    },
  });
  await expect(roundtripResult).toMatchFaviconsSnapshot();
});
