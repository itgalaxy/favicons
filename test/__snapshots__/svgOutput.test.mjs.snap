// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should generate svg favicon 1`] = `
{
  "files": [],
  "html": [
    "<link rel="icon" type="image/svg+xml" href="/favicon.svg">",
  ],
  "images": [
    {
      "contents": "<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1" viewBox="0 0 1024 1024" width="1024" height="1024">
  <image width="1024" height="1024" xlink:href="data:image/"/>
</svg>",
      "name": "favicon.svg",
    },
  ],
}
`;

exports[`should generate svg favicon 10`] = `
{
  "files": [],
  "html": [
    "<link rel="icon" type="image/x-icon" href="/favicon.ico">",
    "<link rel="icon" type="image/png" sizes="16x16" href="/favicon-16x16.png">",
    "<link rel="icon" type="image/png" sizes="32x32" href="/favicon-32x32.png">",
    "<link rel="icon" type="image/png" sizes="48x48" href="/favicon-48x48.png">",
  ],
  "images": [
    {
      "contents": null,
      "name": "favicon.ico",
    },
    {
      "contents": null,
      "name": "favicon-16x16.png",
    },
    {
      "contents": null,
      "name": "favicon-32x32.png",
    },
    {
      "contents": null,
      "name": "favicon-48x48.png",
    },
  ],
}
`;

exports[`should reuse svg as a favicon 1`] = `
{
  "files": [],
  "html": [
    "<link rel="icon" type="image/svg+xml" href="/favicon.svg">",
  ],
  "images": [
    {
      "contents": "<svg xmlns="http://www.w3.org/2000/svg" version="1.1" viewBox="0 0 1500 1500" width="1500" height="1500">
  <path fill="#ff8c8c" fill-rule="evenodd" stroke="#ff8c8c" stroke-linecap="round" stroke-linejoin="round" stroke-width="125" d="M636.39443 914.97056L407.70197 686.75804l-81.36252 82.95151 310.61677 310.46155 537.00668-538.1103-83.356-82.94415z"/>
</svg>
",
      "name": "favicon.svg",
    },
  ],
}
`;
